pkg/ccl/changefeedccl/changefeed_test.go

```diff
@@ -8354,3 +8354,101 @@ func TestChangefeedTopicNames(t *testing.T) {
 
 	cdcTest(t, testFn, feedTestForceSink("pubsub"))
 }
+
+// Regression test for #108450. When a changefeed hits a retryable error
+// and retries, it should start with the most up-to-date highwater (ie. the
+// highwater in the job record). If there is an error reading the highwater
+// from the job record, there should be retries until we are able to get the
+// highwater.
+func TestHighwaterDoesNotRegressOnRetry(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	testFn := func(t *testing.T, s TestServer, f cdctest.TestFeedFactory) {
+		defer changefeedbase.TestingSetDefaultMinCheckpointFrequency(10 * time.Millisecond)()
+		knobs := s.TestingKnobs.
+			DistSQL.(*execinfra.TestingKnobs).
+			Changefeed.(*TestingKnobs)
+
+		sqlDB := sqlutils.MakeSQLRunner(s.DB)
+		sqlDB.Exec(t, `CREATE TABLE foo (a INT PRIMARY KEY)`)
+		foo := feed(t, f, `CREATE CHANGEFEED FOR foo WITH resolved = '10ms'`)
+		defer closeFeed(t, foo)
+
+		// Rough estimate of the statement time. The test only asserts that
+		// things happen after the statement time. Asserting things happen after
+		// this is good enough.
+		initialHighwater := s.Server.Clock().Now()
+
+		jobFeed := foo.(cdctest.EnterpriseTestFeed)
+		jobRegistry := s.Server.JobRegistry().(*jobs.Registry)
+
+		// Pause the changefeed to configure testing knobs which need the job ID.
+		require.NoError(t, jobFeed.Pause())
+
+		// A flag we toggle on to put the changefeed in a retrying state.
+		var changefeedIsRetrying atomic.Bool
+		knobs.RaiseRetryableError = func() error {
+			if changefeedIsRetrying.Load() {
+				return errors.New("test retryable error")
+			}
+			return nil
+		}
+
+		doneCh := make(chan struct{}, 1)
+		knobs.StartDistChangefeedInitialHighwater = func(retryHighwater hlc.Timestamp) {
+			if changefeedIsRetrying.Load() {
+				progress := loadProgress(t, jobFeed, jobRegistry)
+				progressHighwater := progress.GetHighWater()
+				// Sanity check that the highwater is not nil, meaning that a
+				// highwater timestamp was written to the job record.
+				require.NotNil(t, progressHighwater)
+				// Assert that the retry highwater is equal to the one in the job
+				// record.
+				require.True(t, progressHighwater.Equal(retryHighwater))
```

Comment:
You should change this knob to take ctx (so that you can wait on ctx and doneCh below, with timeout), and you should not use require.XXX
here and instead return an error.

Why, you might ask? failing t on any thread other than this test could produce unexpected results -- i..e slow/stuck tests.
t.Fail() (which is called by require.XXX when assertion fails) calls panic() -- okay, great.  But imagine what happens when that code runs under recover() -- that panic doesn't terminate the test; it merely ends this goroutine (invoked by changefeed processor)... Then you continue to wait for things (like ability to write to doneCh, etc) and that may not work correctly.

Does this issue impact this particular test? I don't know -- and I don't want to know -- I just want to be a bit more defensive and not to use the pattern that's known to cause issues.